sirepo/template/radia.py

```diff
@@ -1086,42 +1096,58 @@ def _undulator_termination_name(index, term_type):
     return f'termination.{term_type}.{index}'
 
 
-def _update_cuboid(b, center, size, segments, material, mat_file, magnetization, rem_mag, color):
-    b.center = sirepo.util.to_comma_delimited_string(center)
-    b.color = color
-    b.magnetization = sirepo.util.to_comma_delimited_string(magnetization)
-    b.remanentMag = rem_mag
-    b.material = material
-    b.materialFile = mat_file
-    b.size = sirepo.util.to_comma_delimited_string(size)
-    b.division = sirepo.util.to_comma_delimited_string(segments)
-    return b
+def _update_cuboid(cuboid, **kwargs):
+    return _update_geom_obj(cuboid, delim_fields=PKDict(segments=[1, 1, 1]), **kwargs)
+
 
+def _update_ell(ell, beam_axis, height_axis, **kwargs):
+    ell = _update_geom_obj(ell, **kwargs)
+    _update_ell_points(ell, beam_axis, height_axis)
+    return ell
 
-def _update_geom_from_undulator(geom, und, beam_axis):
+
+def _update_ell_points(ell, beam_axis, height_axis):
+    w, h, b = _geom_directions(beam_axis, height_axis)
+    ctr = sirepo.util.split_comma_delimited_string(ell.center, float)
+    sz = sirepo.util.split_comma_delimited_string(ell.size, float)
+    c = [numpy.sum(w * ctr), numpy.sum(h * ctr)]
+    s = [numpy.sum(w * sz), numpy.sum(h * sz)]
+
+    # start with arm top, stem left - then reflect across centroid axes as needed
+    ax1 = c[0] - s[0] / 2
+    ax2 = ax1 + s[0]
+    ay1 = c[1] + s[1] / 2
+    ay2 = ay1 - ell.armHeight
+
+    sx1 = c[0] - s[0] / 2
+    sx2 = sx1 + ell.stemWidth
+    sy = c[1] - s[1] / 2
+
+    k = [int(ell.stemPosition), int(ell.armPosition)]
+    ell.points = [
+        [ax1, ay1], [ax2, ay1], [ax2, ay2],
+        [sx2, ay2], [sx2, sy], [sx1, sy],
+        [ax1, ay1]
+    ]
+    ell.points = [[2 * c[i] * k[i] + -1**k[i] * v for (i, v) in enumerate(p)] for p in ell.points]
+
+    # Radia's extrusion expects points in permutation order based on the extrusion
+    # axis (x -> [y, z], y -> [z, x], z -> [x, y])
+    if w.tolist().index(1) != (b.tolist().index(1) + 1) % 3:
+        for p in ell.points:
+            p.reverse()
+
+
+def _update_geom_from_undulator(geom, und, beam_axis, height_axis):
```

Comment:
This method is pretty long. I'd consider splitting it up.

Reply:
That and other cleanup is indeed planned